pkg/controller/merged.go

```diff
@@ -1 +1,170 @@
+	t.lock.Lock()
+	defer t.lock.Unlock()
+	_, ok := t.merged[id]
+	if ok {
+		delete(t.merged, id)
+	} else {
+		for hostileID, friendIDs := range t.merged {
+			delete(friendIDs, id)
+			if len(friendIDs) == 0 {
+				delete(t.merged, hostileID)
+			}
+		}
+	}
+}
+
+func (c *controller) broadcastMerges() {
+	merges := c.scope.Merges(c.coalition)
+	newMerges := make(map[brevity.Group][]*trackfiles.Trackfile)
+	for hostileGroup, friendlies := range merges {
+		newMergedFriendlies := map[uint64]*trackfiles.Trackfile{}
+		for _, hostileID := range hostileGroup.ObjectIDs() {
+			for _, friendly := range friendlies {
+				hostile := c.scope.FindUnit(hostileID)
+				logger := log.
+					With().
+					Str("hostile", hostile.Contact.Name).
+					Uint64("hostileID", hostileID).
+					Str("friendly", friendly.Contact.Name).
+					Uint64("friendID", friendly.Contact.ID).
+					Logger()
+
+				isMerged := c.merges.isMerged(hostileID, friendly.Contact.ID)
+				distance := spatial.Distance(friendly.LastKnown().Point, hostile.LastKnown().Point)
+				enteredMerge := distance < brevity.MergeEntryDistance
+				exitedMerge := distance > brevity.MergeExitDistance
+
+				if !isMerged && enteredMerge {
+					logger.Info().Msg("hostile and friendly merged")
+					c.merges.merge(hostileID, friendly.Contact.ID)
+					newMergedFriendlies[friendly.Contact.ID] = friendly
+				} else if isMerged && exitedMerge {
+					logger.Info().Msg("hostile and friendly exited merge")
+					c.merges.separate(hostileID, friendly.Contact.ID)
+				} else if isMerged {
+					logger.Debug().Msg("omitting units from merge call because they were already merged")
+				}
+			}
+		}
+		if len(newMergedFriendlies) > 0 {
+			newMerges[hostileGroup] = slices.Collect(maps.Values(newMergedFriendlies))
+		}
+	}
+
+	for hostileGroup, friendlies := range newMerges {
+		logger := log.With().Stringer("group", hostileGroup).Logger()
+		call := brevity.MergedCall{Callsigns: make([]string, 0), Group: hostileGroup}
+
+		anyOnFrequency := false
+		for _, friendly := range friendlies {
+			logger.Debug().Str("name", friendly.Contact.Name).Msg("checking if friendly contact is on frequency")
+			isOnFrequency := c.srsClient.IsOnFrequency(friendly.Contact.Name)
+			if !c.threatMonitoringRequiresSRS || isOnFrequency {
+				logger.Debug().Bool("isOnFrequency", isOnFrequency).Msg("friendly contact is on frequency (or threat monitoring does not require SRS)")
+				if callsign, ok := parser.ParsePilotCallsign(friendly.Contact.Name); ok {
+					logger.Debug().Str("callsign", callsign).Msg("adding callsign to merged call")
+					call.Callsigns = append(call.Callsigns, callsign)
+					anyOnFrequency = true
+				}
+			}
+		}
+
+		logger = logger.With().Strs("callsigns", call.Callsigns).Logger()
+
+		if anyOnFrequency {
+			logger.Info().Msg("broadcasting merged call")
+			c.out <- call
+		} else {
+			logger.Debug().Stringer("group", hostileGroup).Msg("skipping merged call because no relevant clients are on frequency")
+		}
```

Comment:
this is a good check, but now I'm wondering if this same check exists everywhere? which makes me think there should be some kind of controller class that does this (or any other similar checks) in one spot.

Reply:
It exists for a small number of calls where:

1. The workflow is based on a timer or trigger rather than an explicit request
2. The call is intended for a specific set of recipients instead of the whole coalition

I think it's currently used for THREAT only? PICTURE and FADED use different logic checking if anyone is on frequency at all

Reply:
Factored out into a helper method